database/pdb.ksy

```diff
@@ -0,0 +1,961 @@
+meta:
+  id: pdb_file
+  title: DeviceSQL database export (probably generated by rekordbox)
+  application: rekordbox
+  file-extension:
+    - pdb
+  license: EPL-1.0
+  endian: le
+
+doc: |
+  This is a relational database format designed to be efficiently used
+  by very low power devices (there were deployments on 16 bit devices
+  with 32K of RAM). Today you are most likely to encounter it within
+  the Pioneer Professional DJ ecosystem, because it is the format that
+  their rekordbox software uses to write USB and SD media which can be
+  mounted in DJ controllers and used to play and mix music.
+
+  It has been reverse-engineered to facilitate sophisticated
+  integrations with light and laser shows, videos, and other musical
+  instruments, by supporting deep knowledge of what is playing and
+  what is coming next through monitoring the network communications of
+  the players.
+
+  The file is divided into fixed-size blocks. The first block has a
+  header that establishes the block size, and lists the tables
+  available in the database, identifying their types and the index of
+  the first of the series of linked pages that make up that table.
+
+  Each table is made up of a series of rows which may be spread across
+  any number of pages. The pages start with a header describing the
+  page and linking to the next page. The rest of the page is used as a
+  heap: rows are scattered around it, and located using an index
+  structure that builds backwards from the end of the page. Each row
+  of a given type has a fixed size structure which links to any
+  variable-sized strings by their offsets within the page.
+
+  As changes are made to the table, some records may become unused,
+  and there may be gaps within the heap that are too small to be used
+  by other data. There is a bit map in the row index that identifies
+  which rows are actually present. Rows that are not present must be
+  ignored: they do not contain valid (or even necessarily well-formed)
+  data.
+
+  The majority of the work in reverse-engineering this format was
+  performed by @henrybetts and @flesniak, for which I am hugely
+  grateful. @GreyCat helped me learn the intricacies (and best
+  practices) of Kaitai far faster than I would have managed on my own.
+
+doc-ref: https://github.com/Deep-Symmetry/crate-digger/blob/master/doc/Analysis.pdf
+
+seq:
+  - contents: [0, 0, 0, 0]
```

Comment:
Is it a signature or what is it?

Reply:
This is unknown.

Reply:
In fact, it may be best to just map it as a 4-byte integer, but this way I will get a parse failure if we see something else there, which is interesting information, but not to my users. So perhaps I will change it.

Reply:
I have also made this change.